add precompiled c/c++ header support to zig build

xxxbxxx opened the pull request:
It gets my c++ project full rebuild time from
```
real	10m13.212s
user	139m23.711s
sys	12m1.938s
```

to 
```
real	4m52.636s
user	64m52.851s
sys	5m59.472s
```

pch are a little bit different from other compile steps:
- like binary object files they are built by the C compiler from a (single) source header file.
- however, unlike binary object files, they are not to be fed to the linker, but  as a companion source file to subsequent C compiles.

xxxbxxx commented:
ah. there's a problem still.

somehow didn't notice before but:
 when _not_ cleaning the cache and doing a "git pull && build "

I sometimes get this error
```
:1:1: error: file '/path/to/some_include.h' has been modified since the precompiled header '..../zig-cache/o/1b1f7758f51d4e159b4e5605c9750412/all.pch' was built: mtime changed (was 1700661825, now 1700815558)
:1:1: note: please rebuild precompiled header '..../zig-cache/o/1b1f7758f51d4e159b4e5605c9750412/all.pch'
```

with some_include.h contents unchanged but probably touched by git or something.

so it needs some adjustment to do either to make llvm less strict, or the zig caching system more strict to accommodate for this.

xxxbxxx commented:
> so it needs some adjustment to do either to make llvm less strict, or the zig caching system more strict to accommodate for this.

There is indeed an option for this:
https://releases.llvm.org/17.0.1/tools/clang/docs/ClangCommandLineReference.html#cmdoption-clang-fpch-validate-input-files-content

andrewrk commented:
A few points on this topic:

1. Nice work. I'm willing to merge & maintain this enhancement if you rebase against the latest master and deal with those conflicts.
2. Is the example from the new standalone test that you added a good representation of how this is intended to be used? If so, I want to run another idea by you: perhaps the build system could do this automatically? Is there any downside? What's stopping the build system from always precompiling headers, caching them, re-running that process if they are invalidated, and always passing pch files to clang?

xxxbxxx commented:
>     2. Is the example from the new standalone test that you added a good representation of how this is intended to be used? If so, I want to run another idea by you: perhaps the build system could do this automatically? Is there any downside? What's stopping the build system from always precompiling headers, caching them, re-running that process if they are invalidated, and always passing pch files to clang?

I don't think this would work.
1. making a unique pch for each compilation unit would be slower, so we'd need to automatically aggreagate includes of multiple c files, to be able to reuse the same pch multiple times
2. it is sensitive to include order, #defines, ... so it would be tricky to automate this, I think.
3. you may not want to include every include of the whole program into a single pch, because it turns every touch and build (of a single header) into full rebuild.

What may work is to automatically make a pch header for system/libc/libc++ includes and provide that if none are manually specified. And that could be in the global cache probably.

andrewrk commented:
Interesting. Thanks for that insight. We can consider that idea regarding libc/libc++ in a separate proposal.

xxxbxxx commented:
As an other test, I've added pch support to 
https://github.com/andrewrk/libchromaprint

<details>
<summary> patch </summary>

```diff
diff --git a/build.zig b/build.zig
index e7d4f8e..3774c80 100644
--- a/build.zig
+++ b/build.zig
@@ -18,7 +18,7 @@ pub fn build(b: *std.Build) void {
     lib.linkLibC();
     lib.linkLibCpp();
     lib.addIncludePath(.{ .path = "src" });
-    lib.addConfigHeader(b.addConfigHeader(.{
+    const config_h = b.addConfigHeader(.{
         .style = .{ .cmake = .{ .path = "config.h.in" } },
     }, .{
         .HAVE_ROUND = 1,
@@ -35,7 +35,31 @@ pub fn build(b: *std.Build) void {
         .USE_FFTW3F = null,
         .USE_VDSP = null,
         .USE_KISSFFT = null,
-    }));
+    });
+    const c_flags = &.{
+        "-std=c++11",
+        "-fno-rtti",
+        "-fno-exceptions",
+        "-DHAVE_CONFIG_H",
+        "-D_SCL_SECURE_NO_WARNINGS",
+        "-D__STDC_LIMIT_MACROS",
+        "-D__STDC_CONSTANT_MACROS",
+        "-DCHROMAPRINT_NODLL",
+    };
+    lib.addConfigHeader(config_h);
+
+    const pch = b.addPrecompiledCHeader(.{
+        .name = "chromaprint",
+        .target = target,
+        .optimize = optimize,
+        .cpp_header = true,
+    }, .{
+        .file = .{ .path = "src/pch.h" },
+        .flags = c_flags,
+    });
+    pch.addIncludePath(.{ .path = "src" });
+    pch.addConfigHeader(config_h);
+
     lib.addCSourceFiles(.{
         .files = &.{
             "src/audio_processor.cpp",
@@ -57,16 +81,8 @@ pub fn build(b: *std.Build) void {
             "src/chromaprint.cpp",
             "src/fft_lib_avfft.cpp",
         },
-        .flags = &.{
-            "-std=c++11",
-            "-fno-rtti",
-            "-fno-exceptions",
-            "-DHAVE_CONFIG_H",
-            "-D_SCL_SECURE_NO_WARNINGS",
-            "-D__STDC_LIMIT_MACROS",
-            "-D__STDC_CONSTANT_MACROS",
-            "-DCHROMAPRINT_NODLL",
-        },
+        .flags = c_flags,
+        .precompiled_header = pch,
     });
     lib.addCSourceFiles(.{
         .files = &.{
diff --git a/src/pch.h b/src/pch.h
new file mode 100644
index 0000000..489a7f3
--- /dev/null
+++ b/src/pch.h
@@ -0,0 +1,23 @@
+#include <stdint.h>
+#include <math.h>
+#include <assert.h>
+#include <stdio.h>
+#include <string.h>
+
+#include <vector>
+#include <limits>
+#include <cmath>
+#include <limits>
+#include <algorithm>
+#include <string>
+#include <cstring>
+#include <memory>
+#include <cassert>
+
+extern "C" {
+#include <libavcodec/avcodec.h>
+#include <libavcodec/avfft.h>
+#include <libavutil/mem.h>
+}
+
``` 
</details>


rebuilding only libchromaprint (and not the dependencies) I get something like:

without `.precompiled_header = pch,`:
> real	0m10.904s
> user	0m34.543s
> sys	0m6.052s

with `.precompiled_header = pch,`:
> real	0m10.169s
> user	0m18.176s
> sys	0m4.783s

xxxbxxx commented:
however there's still something going wrong somewhere.

continuing on the libchromaprint example, 
if I:
 - make some change to `src/utils/gaussian_filter.h` (like adding whitespace)
 - then touch `src/pch.h`
 - `zig build` 

```
:1:1: error: file '.../libchromaprint/src/pch.h' has been modified since the precompiled header '.../libchromaprint/zig-cache/o/0a3b415db50f4a754a9af6536bf2a496/chromaprint.pch' was built: mtime changed (was 1705833127, now 1705837082)
:1:1: note: please rebuild precompiled header '.../libchromaprint/zig-cache/o/0a3b415db50f4a754a9af6536bf2a496/chromaprint.pch'
```

the modified file `src/utils/gaussian_filter.h`  is not included by (or any dependency of) pch.h  (it is actually only included by a single cpp file)
So I really don't understand what is going on.

xxxbxxx commented:
ok.
 apparently it was just that `-fpch-validate-input-files-content`  is needed both when using and when emitting the pch.

andrewrk commented on lib/std/Build.zig:
```diff
@@ -714,6 +714,35 @@ pub fn addObject(b: *Build, options: ObjectOptions) *Step.Compile {
+    target: ResolvedTarget,
+    optimize: std.builtin.OptimizeMode,
+    max_rss: usize = 0,
+    error_tracing: ?bool = null,
```
pretty sure this has no effect for a pre-compiled header build.

andrewrk commented on lib/std/Build.zig:
```diff
@@ -714,6 +714,35 @@ pub fn addObject(b: *Build, options: ObjectOptions) *Step.Compile {
+    optimize: std.builtin.OptimizeMode,
+    max_rss: usize = 0,
+    error_tracing: ?bool = null,
+    cpp_header: bool = false,
```
why is this needed?

andrewrk commented on lib/std/Build.zig:
```diff
@@ -714,6 +714,35 @@ pub fn addObject(b: *Build, options: ObjectOptions) *Step.Compile {
+        .root_module = .{
+            .target = options.target,
+            .optimize = options.optimize,
+            .link_libc = !options.cpp_header,
+            .link_libcpp = options.cpp_header,
```
This logic looks wrong. libc++ depends on libc, so link_libcpp implies link_libc.

andrewrk commented on lib/std/Build.zig:
```diff
@@ -714,6 +714,35 @@ pub fn addObject(b: *Build, options: ObjectOptions) *Step.Compile {
+        .max_rss = options.max_rss,
+        .use_llvm = true,
+    });
+    pch.is_linking_libc = !options.cpp_header;
+    pch.is_linking_libcpp = options.cpp_header;
```
These are supposed to be computed during make(), not set during build.zig script logic.

andrewrk commented on lib/std/Build.zig:
```diff
@@ -714,6 +714,35 @@ pub fn addObject(b: *Build, options: ObjectOptions) *Step.Compile {
+        },
+        .kind = .pch,
+        .max_rss = options.max_rss,
+        .use_llvm = true,
```
This can be left as default, since the compiler knows whether it needs to default to using LLVM for outputting a precompiled header.

andrewrk commented on lib/std/Build/Module.zig:
```diff
@@ -84,16 +84,19 @@ pub const CSourceFiles = struct {
     /// else relative to the build root.
     files: []const []const u8,
     flags: []const []const u8,
+    precompiled_header: ?*Step.Compile = null,
```
This should be a `LazyPath` rather than a `Step.Compile`. The pre-compiled header could come from any external process, or even committed directly to source control.

andrewrk commented on lib/std/Build/Step/Compile.zig:
```diff
@@ -263,27 +264,32 @@ pub fn create(owner: *std.Build, options: Options) *Compile {
             .exe => "zig build-exe",
             .lib => "zig build-lib",
             .obj => "zig build-obj",
+            .pch => "zig build-pch",
```
don't forget to add the corresponding usage text in src/main.zig

andrewrk commented on lib/std/Build/Step/Compile.zig:
```diff
@@ -915,7 +927,7 @@ fn make(step: *Step, prog_node: *std.Progress.Node) !void {
         .lib => "build-lib",
         .exe => "build-exe",
-        .obj => "build-obj",
+        .obj, .pch => "build-obj",
```
if it's building a precompiled header, it's not exactly building an object is it?

andrewrk commented on lib/std/Build/Step/Compile.zig:
```diff
@@ -1140,17 +1161,28 @@ fn make(step: *Step, prog_node: *std.Progress.Node) !void {
                         }
+
+                        if (self.kind == .pch) {
+                            try zig_args.append("-x");
+                            try zig_args.append(if (self.is_linking_libcpp) "c++-header" else "c-header");
```
is this controlling the input language or the output language? In either case, whether libc++ is being linked is not the correct logic.

Shouldn't this depend on the file extension of the input header instead?

andrewrk commented on src/main.zig:
```diff
@@ -1417,6 +1417,12 @@ fn buildOutputType(
                         emit_bin = .{ .yes = arg["-femit-bin=".len..] };
                     } else if (mem.eql(u8, arg, "-fno-emit-bin")) {
                         emit_bin = .no;
+                    } else if (mem.eql(u8, arg, "-femit-pch")) {
+                        clang_preprocessor_mode = .pch;
+                        emit_bin = .yes_default_path;
+                    } else if (mem.startsWith(u8, arg, "-femit-pch=")) {
+                        clang_preprocessor_mode = .pch;
+                        emit_bin = .{ .yes = arg["-femit-pch=".len..] };
```
Implementing it this way means that it's possible to write something like this:

```
zig build-exe ... -femit-pch
```

which is nonsense. Right? `zig build-pch` is needed to implement this correctly.

xxxbxxx commented:
yeah I was  trying to sneak in the feature without add new explicit options, for some reason - probably because precompiled header are specific to c/c++ ,  not a zig language thing, 

I tried to modify the change to be more straight forward, that should address all review comments.

andrewrk commented:
I'm sorry, I have let this sit too long unreviewed and acquire conflicts. I will fix them for you and attempt to merge this, but I will first give priority to the other PRs that are closer to a mergeable state.

xxxbxxx commented:
>  acquire conflicts. I will fix them for you 

No trouble: I do use the branch locally and keep it up to date. I just don't push everytime not to waste CI time.. :)

xxxbxxx commented:
I tried to hack together an example with https://github.com/allyourcodebase/cpython :

You can see it is a bit clumsy to use:
-   the options need to be repeated.  (however I don't see how it can be avoided. for example, adding some mechanism to inherit options from the `exe` doesn't make sense because the same pch file can be used on multiple exes)
- it needs some manual setup, since the include context must be taken into account. see NEEDS_PY_IDENTIFIER
- `linkLibC()` is a bit strange since no linking takes place, but it is needed to add the libc include paths.

However I do observe a small speedup:

|              | with pch     | baseline      |
|------------|-----------------|------------------|
|real	| 0m29,353s | 0m42,753s |
|user	| 4m54,776s | 7m50,943s |
|sys	| 1m20,702s | 1m48,318s |

time to rebuild 5 times the cpython source - not the openssl dependency. (done by hacking a fake config.h option that I change to force a rebuild)



```diff
diff --git a/build.zig b/build.zig
index c7fc831039..288e078b96 100644
--- a/build.zig
+++ b/build.zig
@@ -30,7 +30,7 @@ pub fn build(b: *std.Build) void {
     exe.rdynamic = true;
 
     @setEvalBranchQuota(10000);
-    exe.addConfigHeader(b.addConfigHeader(.{
+    const config_header = b.addConfigHeader(.{
         .style = .{ .autoconf = b.path("pyconfig.h.in") },
         .include_path = "pyconfig.h",
     }, .{
@@ -628,7 +628,24 @@ pub fn build(b: *std.Build) void {
         .socklen_t = null,
         .uid_t = null,
         .WORDS_BIGENDIAN = null,
-    }));
+    });
+    exe.addConfigHeader(config_header);
+
+    const precompiled_header = b.addPrecompiledCHeader(.{
+        .name = "Python",
+        .target = target,
+        .optimize = optimize,
+    }, .{ .file = b.path("Include/Python.h"), .flags = &.{
+        "-fwrapv",
+        "-std=c11",
+        "-fvisibility=hidden",
+        "-DPy_BUILD_CORE",
+    } });
+    precompiled_header.linkLibC(); // setup libc include path
+    precompiled_header.addIncludePath(b.path("Include/internal"));
+    precompiled_header.addIncludePath(b.path("."));
+    precompiled_header.addIncludePath(b.path("Include"));
+    precompiled_header.addConfigHeader(config_header);
 
     exe.addCSourceFiles(.{ .files = &.{
         "Modules/_abc.c",
@@ -664,7 +681,6 @@ pub fn build(b: *std.Build) void {
         "Modules/getbuildinfo.c",
         "Modules/itertoolsmodule.c",
         "Modules/main.c",
-        "Modules/mathmodule.c",
         "Modules/md5module.c",
         "Modules/sha1module.c",
         "Modules/sha256module.c",
@@ -805,7 +821,19 @@ pub fn build(b: *std.Build) void {
         "-std=c11",
         "-fvisibility=hidden",
         "-DPy_BUILD_CORE",
+    }, .precompiled_header = precompiled_header.getEmittedBin() });
+
+    // compile without precompiled header:
+    //   the file includes Python.h with NEEDS_PY_IDENTIFIER defined.
+    exe.addCSourceFiles(.{ .files = &.{
+        "Modules/mathmodule.c",
+    }, .flags = &.{
+        "-fwrapv",
+        "-std=c11",
+        "-fvisibility=hidden",
+        "-DPy_BUILD_CORE",
     } });
+
     exe.addCSourceFiles(.{ .files = &.{
         "Modules/getpath.c",
     }, .flags = &.{
```

xxxbxxx commented:
(merged back from https://github.com/ziglang/zig/pull/20879)

I included  a simpler interface to use precompiled headers:

It is now possible to simply add the name of the header file to precompile:
```zig
exe.addCSourceFiles(.{
            .files = &.{"file.c", ...},
            .flags = ...,
            .precompiled_header = .{ .source_header = .{.path = b.path("all.h")} },
        });
```

In order to do this, 
 * I added an extra `finalize()` function to the build steps, 
 * created new build steps not explicitly requested by the user, 
 * as well as altered the compile steps after the user as initialised them. 
I don't know if that fits the general design of the build system or if there's an other way to achieve this.

---

updated example patch for https://github.com/allyourcodebase/cpython:
```diff
--- a/build.zig
+++ b/build.zig
@@ -664,7 +664,6 @@ pub fn build(b: *std.Build) void {
         "Modules/getbuildinfo.c",
         "Modules/itertoolsmodule.c",
         "Modules/main.c",
-        "Modules/mathmodule.c",
         "Modules/md5module.c",
         "Modules/sha1module.c",
         "Modules/sha256module.c",
@@ -805,7 +804,19 @@ pub fn build(b: *std.Build) void {
         "-std=c11",
         "-fvisibility=hidden",
         "-DPy_BUILD_CORE",
-    } });
+    }, .precompiled_header = .{ .source_header = .{ .path = b.path("Include/Python.h") } } });
+
+    // files that includes Python.h with NEEDS_PY_IDENTIFIER defined
+    exe.addCSourceFiles(.{ .files = &.{
+        "Modules/mathmodule.c",
+    }, .flags = &.{
+        "-fwrapv",
+        "-std=c11",
+        "-fvisibility=hidden",
+        "-DPy_BUILD_CORE",
+        "-DNEEDS_PY_IDENTIFIER",
+    }, .precompiled_header = .{ .source_header = .{ .path = b.path("Include/Python.h") } } });
+
     exe.addCSourceFiles(.{ .files = &.{
         "Modules/getpath.c",
     }, .flags = &.{
```

Time to `zig build` when altering `Include/Python.h`:
before:
> real	0m8,841s
> user	1m33,561s
> sys	0m21,765s

after: 
> real	0m6,969s
> user	1m6,623s
> sys	0m17,996s

xxxbxxx commented on lib/std/Build/Module.zig:
```diff
@@ -46,11 +46,36 @@ pub const RPath = union(enum) {
 };
 
+// subset of Compilation.FileExt
+pub const AsmSourceLang = enum {
```
I guess so, but I never used `addAssemblyFile()` myself so I don't know how it is used by people. And I don't know if there is a reason to have it separate from `addCSourceFile()` - which coincidentally also allows for asm files with defines and all.
maybe it should just be removed? 

Altough I don't want to include unrelated changes to the pull request,  looks like I still felt compelled to add `AsmSourceFile` to make the API of `addAssemblyFile()` more similar to `addCSourceFile()` for some reason, maybe I shouldn't have or I should go all the way.   (but that looks like wasted effort if it is to be deleted)

xxxbxxx commented on lib/std/Build/Module.zig:
```diff
@@ -78,21 +103,67 @@ pub const SystemLib = struct {
+    mm,
+    /// "objective-c++-header"
+    hmm,
+    /// "assembler"
+    assembly,
+    /// "assembler-with-cpp"
+    assembly_with_cpp,
```
yeah, as I said I'm more drawn to removing it. and maybe finding a better name for  `addCSourceFile` to reflect  that it is not "C" source, but "not zig" source?

xxxbxxx commented on lib/std/Build/Step/Compile.zig:
```diff
@@ -802,17 +809,20 @@ pub fn linkFrameworkWeak(c: *Compile, name: []const u8) void {
 }
 
 pub fn addCSourceFile(compile: *Compile, source: Module.CSourceFile) void {
+    assert(compile.kind != .pch or compile.root_module.link_objects.items.len == 0); // pch can only be generated from a single C header file
```
it was for adding a header file to a "pch build compile step" but as you pointed out, it's now clearer with the specific `addPrecompiledCHeader`.

(and having to add the assert is a bit odd/annoying, but way less that duplicating almost all CompileStep logic to make a specific "CompilePCHStep")

xxxbxxx commented on lib/std/Build.zig:
```diff
@@ -751,6 +751,31 @@ pub fn addObject(b: *Build, options: ObjectOptions) *Step.Compile {
+        .kind = .pch,
+        .max_rss = options.max_rss,
+    });
+    pch.addCSourceFile(source);
```
agreed. left over of previous versions.

xxxbxxx commented on lib/std/Build/Step/Compile.zig:
```diff
@@ -1113,6 +1114,12 @@ fn getZigArgs(compile: *Compile, fuzz: bool) ![][]const u8 {
 
             // Inherit dependencies on system libraries and static libraries.
             for (dep.module.link_objects.items) |link_object| {
+                if (prev_has_xflag and link_object != .c_source_file and link_object != .c_source_files and link_object != .assembly_file) {
+                    try zig_args.append("-x");
+                    try zig_args.append("none");
+                    prev_has_xflag = false;
+                }
+
```
mmm I don't remember if I was cautious with having a left over "-x option" propagating the rest of the command line, or if there is a actual problem by removing it.
will look into it.

GalaxyShard replied:
src/main.zig has some relevant information on how the flag is parsed

xxxbxxx replied:
yeah I reckon if you mix library/objects files and sources
this line `.positional => switch (file_ext orelse Compilation.classifyFileExt(mem.sliceTo(it.only_arg, 0)))` will propage the -x to unrelated files.
I'll try to make a test case.

(btw, thanks for the review, sorry for the delay)

GalaxyShard commented on lib/std/Build/Step/Compile.zig:
```diff
@@ -817,17 +824,20 @@ pub fn linkFrameworkWeak(c: *Compile, name: []const u8) void {
 
 /// Handy when you have many C/C++ source files and want them all to have the same flags.
 pub fn addCSourceFiles(compile: *Compile, options: Module.AddCSourceFilesOptions) void {
+    assert(compile.kind != .pch); // pch can only be generated from a single C header file
```
I feel it's a bit odd for `compile.addCSourceFiles` (and the similar functions below) to have different behavior than `compile.root_module.addCSourceFiles`, especially considering the latter is invalid as well and would probably crash on the assert in finalize. Additionally, it seems most of the functions in `Step.Compile` don't make much sense for a precompiled header. It may be worth just making a seperate `Step.Compile.Pch` or `Step.PrecompileHeader` similar to what `Step.TranslateC` does.

xxxbxxx replied:
> Additionally, it seems most of the functions in Step.Compile don't make much sense for a precompiled header.

mmm I don't  _most_ .  The precompiled header are sensitive to most compiler options, so for instance even "linkXXX" functions may change compile flags and defines that impact the header compilation even though the actual linking doesn't take place.

GalaxyShard replied:
I wouldn't expect any linker options to affect a precompiled header since linking would be done by the executable/library Compile step, although you might know better than I. Also, I can't think of much use for the functions below:
```
installHeader*
addObjCopy
checkObject
setLinkerScript, setVersionScript // only for the linker
forceUndefinedSymbol // for the linker, doesn't really make sense for a pch
isDynamicLibrary/isStaticLibrary/isDll/producesPdbFile/producesImplib // always false
link* // shouldn't change the output, unless pch's fail to compile when the symbols aren't resolved.
addCSourceFile* // can't compile anything more than the pch
addWin32ResourceFile
setVerboseLink // no linking takes place, therefore no output from the linker
getEmitted* //  most of these would fail assertions on a pch
addObjectFile/addObject
```
That's a large amount of the user-facing API of Step.Compile which wouldn't make sense to use in this context, I think it's worth having a separate step so that the API makes it more clear on what affects the compilation of the PCH, what is impossible/doesn't make sense, and what doesn't affect compilation.

andrewrk commented:
Closing old draft.
